Asana integration

zanemcca opened the pull request:
## Completed
- Add Oauth login with Asana
- Pull Asana tasks from asana
- Create UI/UX to create an Asana task
- Persist asana task to Asana
- Created a preview environment at https://tp4000preview.zanemccaig.com/

## Known Issues
https://github.com/zanemcca/task_planner_4000/issues/3

## Preview
Live: https://tp4000preview.zanemccaig.com/
https://www.loom.com/share/e74b6e0b55db48cda5e96c8de96a95c7

zanemcca committed:
Added an initial asana task creation form

zanemcca committed:
Added preview deployments

zanemcca committed:
Added env vars to cloud run deployment

zanemcca committed:
Fixed workflows

zanemcca committed:
Saving current asana integration

zanemcca committed:
Trying to fetch the url instead of pushing it

zanemcca committed:
Added a basic server to handle oauth callbacks

zanemcca committed:
Got asana oauth finally working

zanemcca committed:
Cleaned up some stuff

zanemcca committed:
Updated the readme.md

zanemcca committed:
Added asana token detection in asana task menu

zanemcca committed:
Fixed lint issues

zanemcca committed:
Fixed dockerfile and server

zanemcca committed:
Fixed yarn error

zanemcca committed:
Got asana tasks loading

zanemcca committed:
Got asana task creation working

zanemcca committed:
Got asana integration working with a stateful hook

zanemcca committed:
Got asana task creation syncing with the UI

zanemcca committed:
Cleaned up some state management and lint issues

zanemcca committed:
Fixed the order of task after creation

zanemcca committed:
Got focus working properly

zanemcca committed:
Added env vars to the github flows

zanemcca committed:
Updated icons and title

zanemcca committed:
Added some logs to help debug prod issues

zanemcca committed:
Fixed the https mismatch on prod

zanemcca committed:
Added explicit NODE_ENV=production to dockerfile

zanemcca committed:
Added a log to double check env vars

zanemcca committed:
Trying a couple options for getting cloud run vars working

zanemcca committed:
Another fix for env var injection

zanemcca committed:
Another try

zanemcca committed:
Another fix for the env vars

zanemcca committed:
Fixed prod deployment envs

zanemcca committed:
Removed logs from server

zanemcca commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add Asana Task Creation Form](https://trello.com/c/yGXjXMKg/3-add-asana-task-creation-form)

zanemcca commented:
Just a heads up my next PR is already up and it uses these changes as a basis so the next PR should be merged after this one.

Also I have updated the preview URL to now point to the latest changed from the next PR so there are more features available in the preview. If I were working on this in production I would try and setup a per PR based preview env rather than a shared one to eliminate this issue.

priyadarshy commented on package.json:
```diff
@@ -3,23 +3,29 @@
-    "typescript": "~3.7.2"
+    "react-use": "^13.24.0",
+    "typescript": "~3.7.2",
+    "use-persisted-state": "^0.3.0"
```
This is pretty cool. I didn't know this existed, in most of our web app we use a similar pattern via Meteor's "Reactive Var" and "Session" so that we can achieve this. We could use this to replicate the `jiraState` pattern with react hooks without any provider gymnastics like with `useContext`. cc @bone-saw @BrunoBernardino

zanemcca replied:
It is a pretty convenient package. I just discovered it myself actually. I am finding that almost everything has a nice hook based approach now and I am loving it.

BrunoBernardino replied:
I just browsed through their source code, there's no magic involved and I liked it! It's basically a cool "wrapper" around `localStorage` mapped to `useState`.

priyadarshy commented on src/components/Commands.tsx:
```diff
@@ -48,8 +41,63 @@ const Command = (props: any) => {
+        setIsAsanaTaskFormVisible(true)
+      }
+    }
+  }]
```
This is a really nice touch in the case where a user has yet to add the Asana integration. Great work!

priyadarshy commented on src/components/Commands.tsx:
```diff
@@ -48,8 +41,63 @@ const Command = (props: any) => {
+        return;
+      }
+
+      await createTask({
+        description: values.title
+      })
```
It's likely that customers would want the task creation experience to feel fast. 

Instead of waiting for the task to create here it would be a better experience to kickoff the task creation process and show an alert if it were to fail. 

It might even be nice to show the "loading" dialog in the Asana panel instead of the in the task creation form so that the user can return to whatever they were doing before and have the feedback that the app was writing something to Asana.

zanemcca replied:
That is a good point.

I thought about going through the process of adding in error recovery and cleanup with a fake it first approach but I found the current lag to be reasonable. I agree with you though that moving the loading to the asana task list would probably be a smoother experience. I was thinking a custom CardLoading component with some entry animations like these ones https://cssanimation.rocks/list-items/ might be a nice touch.

Thanks for the feedback in a production environment I would definitely implement these changes.

priyadarshy commented:
Nice work here! 

Overall, It's pretty nice to be able to create tasks without leaving the keyboard! I like that the command palette provides an option to link Asana if I don't have it already. I can see how this could be quickly and easily extended to someday provide a customer the option to select a project and other fields when creating the task. It's pretty neat to see the task just show up in the Asana panel after hitting submit. 

There's a couple of small improvements I think that would make this feel smoother:
- Hitting enter with the modal open can just submit the the form. It's a lot of extra work to tab through the cancel button in order to hit enter to create when that's the primary action. 
- We probably don't want to block the user's experience on waiting for the task to be created in Asana. Instead of showing the loading in the task creation form/modal, I'd return immediately and show an animation in the Asana panel itself letting the user know their task is being created. Or just optimistically add the task and just throw an "oops something went wrong alert" in the rare situation that something breaks. 
- Continue to show the Asana logo somewhere in the task creation form so the user continues to have context and confidence (visually, without reading) that they're about to make a change in another tool. 

I found one small UI bug. If you try and submit an Asana task without a title you get two warnings. I think the "Toast" notification above the modal can probably just be removed. 

![Screen Shot 2020-02-09 at 2 18 03 PM](https://user-images.githubusercontent.com/1954486/74111237-1af1bc80-4b48-11ea-964b-a814ba1cb48d.png)

priyadarshy commented:
I've looked over the code, and tried it in the preview environment (knowing full well that it includes other changes). 

I'll go ahead and merge this in. Don't feel like you need to make any of the changes I mentioned in my review. I just wanted to leave a review exactly as though this was being made in our actual codebase :)

zanemcca commented:
- **"Hitting enter with the modal open can just submit the the form. It's a lot of extra work to tab through the cancel button in order to hit enter to create when that's the primary action."**
  - I actually started to implement it that way using [mousetrap](https://github.com/ccampbell/mousetrap). What it came down to was that the UI form components all supported tab to go to the next component and enter to submit the form's input. This meant that overriding the enter command resulted in some inputs being ignored. I made this as a compromise solution that allowed me to move faster. I completely agree that tabbing through everything is a little annoying though. 👍  on supporting enter to submit.

- **"We probably don't want to block the user's experience on waiting for the task to be created in Asana. Instead of showing the loading in the task creation form/modal, I'd return immediately and show an animation in the Asana panel itself letting the user know their task is being created. Or just optimistically add the task and just throw an "oops something went wrong alert" in the rare situation that something breaks."**
  - I've addressed this one in a reply above. 👍 agree with this one as well.

- **"Continue to show the Asana logo somewhere in the task creation form so the user continues to have context and confidence (visually, without reading) that they're about to make a change in another tool."**
  - I like this a lot. 👍  to add it.

- **"I found one small UI bug. If you try and submit an Asana task without a title you get two warnings. I think the "Toast" notification above the modal can probably just be removed."**
  - Nice catch! In a prod environment I would have more tests in place that hopefully would have caught that. I think it was a regression that I added in after I completed the asana task creation flow. On that note I ordinarily would have written tests for most components and functions in the code in order to maximize functional code coverage. Also since UI/UX is a top priority I would likely have a lot more snapshot tests. I went pretty light on tests for the last few PRs as per @BrunoBernardino tip that they were not really necessary for this project.

BrunoBernardino commented on README.md:
```diff
@@ -37,8 +41,7 @@ Instead, it will copy all the configuration files and the transitive dependencie
+## Developers
 
-To learn React, check out the [React documentation](https://reactjs.org/).
+### Asana Integration
```
Thanks for the extra instructions, here!

zanemcca replied:
No worries. They are still not ideal but I figured for this context it would be enough to get it working on your machine. Ideally I would reconfigure things so that starting it up with the backend is still one command

BrunoBernardino commented on server.js:
```diff
@@ -0,0 +1,57 @@
+ *   [export PORT=...]
+ *   node oauth_webserver.js
+ */
+var Asana = require('asana');
```
I'm curious on the inconsistencies here with `var` vs `const`?

zanemcca replied:
Most of this code is copy and paste from the Asana examples which are in js. Rather than deal with integrating it into the typescript and reworking the package.json scripts I opted for the quick and dirty solution of leaving the server as a js server that just serves the static build  items. I did not bother refactoring to const because without the ts compiler on-top there might be a risk that ES6 would not be supported by my/your node versions and this was already working.

I also went on a bit of a tangent to get asana auth working. Their docs about the auth flow are a little outdated. The frontend "implicit grant flow" is no longer supported so this was all added to allow a proper backend flow.

zanemcca replied:
In a production environment I would ideally move the `server.js` into `src/server.ts` and replace the `yarn start` script with one that will support the frontend and backend.

BrunoBernardino replied:
Thanks for the answer and context!

BrunoBernardino commented on src/components/Asana.tsx:
```diff
@@ -1,37 +1,41 @@
+        {loading && <Row className="Asana-loading">
+          <Spin />
+        </Row>}
+        {map(tasks, task => (
```
Is there any specific reason to use `lodash.map` vs `tasks.map` here?

zanemcca replied:
If for some reason `tasks` is `undefined` or `null` then `lodash.map` will happily accept it and return an empty list that will render into nothing where as `tasks.map` would cause an error. In this context `tasks` should not be `undefined` or `null` as I have ensured it is initialized and I have protected it with typescript typings but sometimes things happen at runtime that are not expected.

BrunoBernardino replied:
Thanks for the answer!

BrunoBernardino commented on src/components/AsanaCallback.tsx:
```diff
@@ -0,0 +1,44 @@
+import { ASANA_REDIRECT_URI } from '../lib/asana';
+import { message } from 'antd';
+import CenterSpin from './CenterSpin';
+import { useMount } from 'react-use';
```
TIL about this pretty cool package! Thanks!

zanemcca replied:
It is fantastic. I only recently starting using it and I love all of it.

BrunoBernardino commented on src/hooks/asana.tsx:
```diff
@@ -0,0 +1,157 @@
+          const user = await client.users.me()
+
+          const userId = user.gid;
+          // The user's "default" workspace is the first one in the list, though
```
Nice comment.

priyadarshy replied:
Ha! I was actually going to say the same thing but decided it was going to be weird to call out a "good comment". But now that we're on it, let me pile on :)

BrunoBernardino reviewed (commented):
This looks great, and Ashutosh already made a point on the most important things. I only have a couple of questions. Good job!
